database/engine/rrdengineapi.c

```diff
@@ -1034,3 +1034,264 @@ void rrdeng_prepare_exit(struct rrdengine_instance *ctx)
+    UNUSED(rep_state);
+    struct pg_cache_page_index* page_index;
+    RRDDIM *rd = dim_past_data->rd;
+    page_index = dim_past_data->rd->state->page_index;
+#ifndef NETDATA_INTERNAL_CHECKS
+    UNUSED(rd);
+#endif
+
+    uv_rwlock_wrlock(&page_index->lock);
+    --page_index->writers;
+    uv_rwlock_wrunlock(&page_index->lock);
+    debug(D_REPLICATION, "%s Finalize operation -  Dimension \"%s\".\"%s\" metrics page completed.", REPLICATION_MSG, rd->rrdset->id, rd->id);
+}
+
+//It updates the dim_past_data according to new start and end times
+void modify_dim_past_data(RRDDIM_PAST_DATA *dim_past_data, usec_t start_time, usec_t end_time)
+{
+    uint64_t start, end, new_start, new_end, new_entries;
+    start = dim_past_data->start_time / USEC_PER_SEC; //gap time start
+    end = dim_past_data->end_time / USEC_PER_SEC;     //gap time end
+    new_start = start_time / USEC_PER_SEC;
+    new_end = end_time / USEC_PER_SEC;
+    new_entries = (uint64_t)(new_end - new_start) / dim_past_data->rd->update_every + 1;
+#ifndef NETDATA_INTERNAL_CHECKS
+    UNUSED(end);
+#endif
+
+    dim_past_data->page_length = new_entries * sizeof(storage_number);
+    if(new_start != start) {
+        storage_number *sn = dim_past_data->page;
+        dim_past_data->page = &sn[((new_start - start) / dim_past_data->rd->update_every) * sizeof(storage_number)];
+    }
+    dim_past_data->start_time = new_start * USEC_PER_SEC;
+    dim_past_data->end_time = new_end * USEC_PER_SEC;
+
+    debug(D_REPLICATION,
+        "%s: Divided page %p - [%llu, %llu, %lu, %lu, %u, %lu]",
+        REPLICATION_MSG,
+        dim_past_data->page,
+        dim_past_data->start_time,
+        dim_past_data->end_time,
+        start,
+        end,
+        dim_past_data->page_length,
+        ((new_start - start) / dim_past_data->rd->update_every) * sizeof(storage_number));
+}
+
+// It saves the GAP past metrics in the active page in real-time
+int rrdeng_store_past_metrics_realtime(RRDDIM *rd, RRDDIM_PAST_DATA *dim_past_data)
+{
+    struct rrdeng_collect_handle *handle = (struct rrdeng_collect_handle *)rd->state->handle;
+    struct rrdeng_page_descr *descr;
+    storage_number *page, *page_gap;
+    int return_value = 0;
+
+    descr = handle->descr;
+    if(!descr || !descr->pg_cache_descr) {
+        infoerr("%s: No active descr or page for dimension %s.%s", REPLICATION_MSG, rd->rrdset->id, rd->id);
+        return 1;
+    }
+
+    page = (storage_number *)descr->pg_cache_descr->page;
+    page_gap = (storage_number *)dim_past_data->page;
+
+    uint64_t start, end, page_start, page_end;
+    start = dim_past_data->start_time / USEC_PER_SEC; //gap time start
+    end = dim_past_data->end_time / USEC_PER_SEC;     //gap time end
+    page_start = descr->start_time / USEC_PER_SEC;    //active page time start
+    page_end = descr->end_time / USEC_PER_SEC;        //active page time start
+
+    if (!page || !page_gap || start > end) {
```

Comment:
Division is one of the worse operations that computers can do, considering that you are dividing by a constant all values and testing them after this, I suggest the opposite. It is better firstly to test the values and only if it is possible to do work with the data we execute a division. 

I hope I am not being hard with requests in this review, so if you need to relax, please, do the following exercise that explains what I wrote above: Proof that if A > B then A*C > B*C. Any doubt ping @vlvkobal . :smile: